Refactors ActionListCard to be ActionList component

burkeshartsis opened the pull request:
## DESCRIPTION
This PR is the scaffolding for https://github.com/ApollosProject/apollos-apps/pull/1428 and is a blocker.

This PR forks the `ActionListCard` component and refactors it to be a more flexible `ActionList` component. It cleans up a lot of stuff and I'm thinking we'll deprecate `ActionListCard`. **I'm not 100% confident in the direction this PR went so please consider this an open call for feedback!**

<img src="https://user-images.githubusercontent.com/2053547/79898962-1106f800-83da-11ea-86a7-bb9a7b654389.gif" width="30%" />

### What does this PR do, or why is it needed?
We are starting to introduce a version of this UI element that doesn't render in a card. This PR mainly refactors `ActionListCard` to render as either a card or a straight list. It defaults to `isCard: true` so it is backwards compatible with `ActionListCard`.

This also updates to some new design spacing and other minor style changes.

### What design trade-offs/decisions were made?
I went back and forth on the following possible implementation paths. 

1. Separate but otherwise identical components for rendering in a card and not. 
  - This felt too wet. Even if I shared components between the two it just seemed unnecessary.
2. Single component that doesn't render in a card. When you want in in a card you just wrap it in one.
  - I think I like this idea the most. It's the most flexible but it is also the least backwards compatible option, requiring the most changes for client apps.
3. Single component that supports rendering in a card or not.
  - This is the usual route most people go and the one I generally avoid. Adding additional API to a component is easy but also complicates maintenance and support. It also opens the door to adding more "features/support" down the road, exacerbating the problem further. That said, the actual implementation felt easy and, while I don't have a strong opinion on this yet, if we are going to continue to add otherwise identical ui elements that don't render in a card this might be the best way to support that.

So I ended up going with option 3. **I'm not sold on it or the `isCard` prop and would happily entertain other ideas**. We could also delete `ActionListCard` now and export `ActionList` as both names since it's backwards compatible.

### How do I test this PR?
Check out the stories for the `ActionList` component.

### What automated tests did you write?
Everything should be covered 💥 

## TODO:

- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [tag-issues-here]
- [x] No new warnings in tests, in storybook, and in-app
- [x] Upload GIF(s) of iOS and Android
- [x] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

vinnyjth commented on packages/apollos-ui-kit/src/ActionList/index.js:
```diff
@@ -0,0 +1,103 @@
+    this.props.isCard ? (
+      <Card isLoading={this.props.isLoading}>{children}</Card>
+    ) : (
+      <>{children}</>
```
Why is this fragment needed?

burkeshartsis replied:
This is a fragment so that if a `header` is passed it will render inside something. I didn't actually test if it was necessary but in head the render order would suggested it is.

burkeshartsis replied:
Tested this and you were correct. Removed the fragment in ed96ff8

vinnyjth commented on packages/apollos-ui-kit/src/ActionList/index.js:
```diff
@@ -0,0 +1,103 @@
+      header,
+    } = this.props;
+
+    const RenderAsCard = this.RenderAsCard; // eslint-disable-line prefer-destructuring
```
Any reason you're not destructuring here?

burkeshartsis replied:
You know... I was sort of blanking on the syntax for that for whatever reason since it was a class member. Of course the syntax is the same no matter what 🙃 Fixed in 2a3154b

burkeshartsis replied:
Actually I reverted that last commit because it doesn't work if you use destructuring 🤷‍♂️

burkeshartsis replied:
Ok actually working here 0816caa

vinnyjth commented on packages/apollos-ui-kit/src/ActionList/index.js:
```diff
@@ -0,0 +1,103 @@
+              onPress={() =>
+                onPressActionItem({
+                  action: item.action,
+                  relatedNode: item.relatedNode,
```
Looking good! I might consider spreading `...props` in here. One thing I've discovered is it can be hard to grant this function access to something like the `UserWebView` without.

burkeshartsis replied:
when you say `props` do you mean the component props or the `item` from the `.map()`?

vinnyjth commented on packages/apollos-ui-kit/src/ActionList/ActionListItem/index.js:
```diff
@@ -0,0 +1,69 @@
+}))(ConnectedImage);
+
+// eslint-disable-next-line react/prop-types
+const RenderAsTouchable = ({ children, onPress }) =>
```
I like the style of this Component, but is it necessary? How does `Touchable` handle not having an `onPress` prop? I think it acts just like a normal view, but I could be mistaken.

burkeshartsis replied:
The problem was that `TouchableScale` still animates even if there is not `onPress` function. I agree though that the fix should live in `TouchableScale`! I'm going to open a PR to fix this there! I'll also add a `TODO` here to remove this fix in that PR.

burkeshartsis replied:
I got a working solution in this component in 2f1745b. I'll look to open a long term fix in another PR as noted above.

burkeshartsis commented:
@vinnyjth this is ready for review again. I think I addressed all of your comments.

Do you have any input my the "What design trade-offs/decisions were made" section before you approve this?
